code/raw-trie.lisp

```diff
@@ -0,0 +1,148 @@
+(cl:in-package #:spell)
+
+;;; `raw-node' class
+
+(defclass raw-node (node) ())
+
+(defmethod %insert ((object t) (string string) (suffix (eql 0)) (node raw-node))
+  (change-class node 'raw-leaf-node)
+  (%insert object string 0 node))
+
+(defmethod %insert ((object t) (string string) (suffix integer) (node raw-node))
+  (change-class node 'raw-interior-node)
+  (%insert object string suffix node))
+
+;;; `raw-leaf-mixin' class and leaf node protocol methods
+
+(defclass raw-leaf-mixin (leaf-mixin)
+  ((%entries :type   simple-vector
+             :writer (setf %entries))))
+
+(defmethod utilities.print-items:print-items append ((object raw-leaf-mixin))
+  (let ((entries '()))
+    (map-entries (lambda (entry)
+                   (push (class-name (class-of entry)) entries))
+                 object  (%entries object))
+    `(((:entries (:after :children)) " entries: ~{~A~^ ~}" ,entries))))
+
+(defmethod map-entries
+    ((function function) (node raw-leaf-mixin) (entries vector))
+  (declare (type simple-vector entries))
+  (map nil function entries))
+
+(defmethod add-entry ((entry t) (node raw-leaf-mixin) (entries vector))
+  (declare (type simple-vector entries))
+  (let ((new-entries (make-array (+ 1 (length entries)))))
+    (setf (aref new-entries 0)   entry
+          (subseq new-entries 1) entries)
```

Comment:
Isn't a `vector-push-extend` better than the case of always fully copying the vector? Is it necessary to add the new entry in the front?

Reply:
Initially, I wanted to not waste any space to reduce the peak heap usage during compilation. But `vector-push-extend` with a small `:extension` could work. On the other other hand, an adjustable array might be slower and larger by itself.

Reply:
Right, might need to benchmark that one and check in practice. An extra step that simplifies all arrays after everything is done might (not) be a decent idea, maybe.

Reply:
The transformation to the compact trie already re-creates the entry array, so that would already happen.